robotiq_c_model_control/src/robotiq_c_model_control/gripper_driver.py

```diff
@@ -0,0 +1,255 @@
+        if all_detected:
+            self.__log.info('All grippers detected!')
+            self.comm_channel = comm_channel
+            self.grippers = self.build_grippers(dev_clients)
+            return True
+        else:
+            comm_channel.disconnect()
+            return False
+
+    def _autodetect(self, find_count):
+        return any(self._autodetect_grippers_with_dev(dev, find_count) for dev in self.DEVICES)
+
+    def _autodetect_grippers_with_dev(self, dev_path, find_count):
+        """
+        Try to detect grippers using the USB to RS-485 converter at the
+        specified path.
+        """
+        comm_channel = SingleCommunication()
+        if not comm_channel.connect(dev_path):
+            self.__log.warn('{}: Could not connect'.format(dev_path))
+            return False
+
+        self.__log.info('{}: Comms connected'.format(dev_path))
+
+        # Scan for grippers attached at all dev_ids
+        dev_clients = []
+        for dev_id in VALID_DEVICE_IDS:
+            comm_client = SingleCommClient(dev_id, comm_channel)
+            if self._is_gripper_attached(dev_id, comm_client):
+                dev_clients.append(comm_client)
+                self.__log.info('Autodetected {}'.format(dev_id))
+                find_count -= 1
+                if find_count == 0:
+                    break
+
+        if find_count == 0:
+            self.__log.info('Autodetect successful')
+            self.comm_channel = comm_channel
+            self.grippers = self.build_grippers(dev_clients)
+            return True
+        else:
+            comm_channel.disconnect()
+            return False
+
+    def _start(self):
+        if self.grippers:
+            self.activate(self.grippers)
+
+    def _shutdown(self):
+        for gripper in self.grippers:
+            gripper.shutdown()
+
+        if self.comm_channel is not None:
+            self.comm_channel.disconnect()
+
+
+@ready_logging.logged
+class TeachmateGripperDriver(GripperDriver):
+    """ Driver for running multiple Robotiq grippers on the Teachmate MODBUS. """
+    def __init__(self):
+        super(TeachmateGripperDriver, self).__init__()
+        self.grippers = []
+        self.clients = []
+
+    def _detect_gripper_with_comms(self, dev_id, comms):
+        """
+        Attempt to connect to a gripper using the specified communication
+        channel.
+        """
+        if not comms.connect():
+            self.__log.warn('Device {}: Could not connect'.format(dev_id))
+            return False
+
+        return self._is_gripper_attached(dev_id, comms)
+
+    def _detect(self, modbus_ids):
+        """
+        Attempt to detect all the specified grippers.
+
+            Each gripper communicates using its own unique client.
+        """
+        for dev_id in modbus_ids:
+            comms = Communication(dev_id)
+            self.clients.append(comms)
+            if not self._detect_gripper_with_comms(dev_id, comms):
+                return False
```

Comment:
Wouldn't you want to remove this comms from self.clients?

Reply:
It works currently because even on `_detect` failure `_shutdown` gets called and disconnects all the open comms.

I'll reorganize it so that is no longer required.